### ISSUE TITLE
Simplified Snapshots or Reading-Mode Support

### ISSUE THREAD

**Issue by about-code** [zotero/zotero-connectors#194]
In an older [commit message](https://github.com/zotero/zotero/commit/c27ff75e80dae96c88779e9490eab798989e471a) @dstillman describes the benefits of saving web pages in Firefox's Reading Mode (`about:reader?url=...`) which was possible in *Zotero for Firefox* up to version 4.

The most important points for me:
> - [...] save what people actually care about - the text of the page
> - avoids filling up people's storage directories and storage
accounts with junk

With Zotero 5 and *Zotero Connector* I can no longer save pages in Reading Mode. The connector shows a website-type icon but when clicked seems to do nothing.

While supporting reading mode would be a quick win for Firefox users it is still not the best long-term solution to above issues and for long-living archives. My archived reading-mode pages looked always broken in non-Firefox browsers and today do so even in Firefox - mainly due `reader.css` and `aboutReader.js`. However by ocassionally doing a *search-and-delete* for those files as well as deleting the reader's SVGs I could achieve good and compact results in the past.

So a simplified snapshot mode really should just capture content and provide an HTML with minimal styling for better readability.

**Comment by dstillman** [comment]
I guess we never rolled that out, but since we'll be leaving the Mozilla platform eventually, we should find a suitable Readability-like JS library rather than trying to port Firefox's Reader Mode to Zotero 5 (where snapshot saving happens). I'm not sure what the current options are. If there was something that could either 1) operate on the HTML itself rather than the DOM or 2) generate cleaned HTML for the current page without actually modifying the DOM, we could even avoid a hidden browser in the client altogether and make snapshot saving vastly quicker. (It's possible we'd need to keep full snapshots as a preference, but I'd be happy defaulting to simplified snapshots.)

(And no, I don't think we'd bother supporting saving from Reading Mode in the Firefox connector.)

**Comment by about-code** [comment]
> rather than trying to port Firefox's Reader Mode to Zotero 5

Fully agree. Wouldn't be worth the effort and very vendor-specific.

> It's possible we'd need to keep full snapshots as a preference, but I'd be happy defaulting to simplified snapshots

Full snapshots are still a unique feature of Zotero and shouldn't be dropped, indeed. As far as I see FF's reading mode only works on pages with a semantic `<article>` element. Its a suitable restriction, given that most content management systems support it these days. I can imagine a test for existence of an `<article>` element being implemented into connectors. Then connectors could present the user a selection between two separate capture modes. Which one to make default in presence of an `<article>` should indeed be a preference. 

> we should find a suitable Readability-like JS library

To me such library wouldn't be essential. I wouldn't like to see it baked into snapshots. For me the issue is less about readability in the first place but about a lightweight snapshot solution. For example, I consider it still a drawback that I replicated Firefox's `reader.js` and `reader.css` with every snapshot in the past. A simplified snapshot solution should really just provide a plain HTML template with a few semantic HTML tags and metadata and optionally (not even required) minimal styling to make reading *a bit* more of a pleasure.

I wouldn't like to see JavaScript being embedded because scripts are neither markup nor content, they are turing-complete programmes which always require a JS runtime to be available. This makes them less friendly for long-term archiving, I'd say. Instead the resulting HTML should simply keep the `<article>` tag. Then people could apply a browser's reading mode again when opening the file. 

What's worthwile: a snapshot could include kind of a "Zotero Simplified Snapshot Signature Metadata". Then a Zotero Connector or any other browser extensions could recognize these snapshots and could automatically apply a reading-mode template/theme by means of a web extension. The snapshot itself should be kept very close to the original ideas of Tim Berners Lee and Markup - which didn't really include JavaScript.

> we could even avoid a hidden browser in the client altogether and make snapshot saving vastly quicker

I am a bit sceptical whether snapshots could or should be produced without a hidden browser and DOM. A lot of interactive and multimedia sites these days unfortunately show or insert `<article>` content only after some javascript has been executed. Websites implemented using a *Single Page Application* approach, for example, couldn't be scraped at all without a browser environment. We'd be limiting applicability of simplified captures to only the parts of the web which render their HTML server-side. I think that'd be a big tradeoff. These additional complexity is in fact a remainder what it means that the web has grown beyond *just markup* and why I advocate for a very simple archive-friendly snapshots in the first place.

**Comment by dstillman** [comment]
>>    we should find a suitable Readability-like JS library
>
>To me such library wouldn't be essential. I wouldn't like to see it baked into snapshots.

The JS library wouldn't be required for viewing. We just need something to produce clean HTML that we can save to disk, and that needs to be written in JavaScript. In the linked commit, we were just using Reader Mode to produce that clean HTML — we weren't embedding the Reader Mode JS as you were getting when you were saving from Reader Mode directly.

>> we could even avoid a hidden browser in the client altogether and make snapshot saving vastly quicker
>
> I am a bit sceptical whether snapshots could or should be produced without a hidden browser and DOM

The page is already loaded in the browser, and we have access to the HTML of the current DOM via `.innerHTML`. So we'd just be saving a simplified version based on the generated page — what comes over the wire doesn't matter.

(The one exception here is when saving multiple results. That will still require a hidden browser in the client, since the connector just passes URLs to the client for those.)

**Comment by about-code** [comment]
> The one exception here is when saving multiple results. That will still require a hidden browser in the client, since the connector just passes URLs to the client for those.

Okay, I see. I thought the connector would always send an URL even for a single snapshot and the snapshot  would always be made by loading a page twice in the hidden browser of the client. Based on that assumption trying to avoid the hidden browser didn't make much sense to me in the first moment but now I think I understand how you plan to avoid it.

**Comment by dstillman** [comment]
I'm not sure what other options are out there, but the first thing to try here might extracting Firefox's Reader Mode code and seeing if we can use a modified version of that. The benefit is that's likely to be actively maintained going forward, unlike (I would guess) most other solutions.

**Comment by adomasven** [comment]
This should help https://github.com/mozilla/readability

**Comment by mlinchitstest** [comment]
any plans for pdf snapshots? currently, "Print Friendly and PDF" is best way to generate a faithful simplified pdf of a page.

**Comment by Victor239** [comment]
Yeah this really limits the usability of saving webpage snapshots in Zotero 7. Adding highlights and sticky notes has poor UX when the saved webpage looks like a mess.
